Basket-Donation-Payments/Basket-Donation-Payments/View Controllers/Login & Signup Flow/NonprofitSignupViewController.m

```diff
@@ -61,31 +61,66 @@ -(void)saveNonprofitAndUserToParse {
     self.nonprofit.category = self.nonprofitCategoryTextField.text;
     self.nonprofit.websiteUrlString = self.nonprofitWebsiteTextField.text;
     self.user.nonprofit = self.nonprofit;
-    [self.user signUpInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
-        if (succeeded) {
-            if (self.finishedSavingBoth) {
-                [self performSegueWithIdentifier:@"loginSegue" sender:nil];
-            } else {
-                self.finishedSavingBoth = true;
-            }
+    
+    NSOperationQueue *opQueue = [NSOperationQueue new];
 
-        } else {
-            UIAlertController *alert = [Utils createAlertControllerWithTitle:@"Could not save user." andMessage:error.localizedDescription okCompletion:nil cancelCompletion:nil];
-            [self presentViewController:alert animated:YES completion:nil];
-        }
+    
+    NSBlockOperation *signupUserOp = [NSBlockOperation new];
+    signupUserOp.queuePriority = NSOperationQueuePriorityHigh;
+    __weak NSBlockOperation *weakSignupUserOp = signupUserOp;
+    [signupUserOp addExecutionBlock:^{
+        [self.user signUpInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
+            if (!succeeded) {
+                [opQueue cancelAllOperations];
+                UIAlertController *alert = [Utils createAlertControllerWithTitle:@"Could not save user." andMessage:error.localizedDescription okCompletion:nil cancelCompletion:nil];
+                [self presentViewController:alert animated:YES completion:nil];
+            }
+        }];
     }];
-    [self.nonprofit saveInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
-        if (succeeded) {
-            if (self.finishedSavingBoth) {
-                [self performSegueWithIdentifier:@"loginSegue" sender:nil];
-            } else {
-                self.finishedSavingBoth = true;
+    
+    NSBlockOperation *saveNonprofitOp = [NSBlockOperation new];
+    saveNonprofitOp.queuePriority = NSOperationQueuePriorityHigh;
+    __weak NSBlockOperation *weakSaveNonprofitOp = saveNonprofitOp;
+    [saveNonprofitOp addExecutionBlock:^{
+        [self.user signUpInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
+            if ([weakSaveNonprofitOp isCancelled]) {
+                [self.user deleteInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
```

Comment:
If this is too complicated, I think you can also serialize these queries (i.e save the user first and then save the nonprofit only if the user is saved successfully). My intuition is that both should run pretty fast and it wouldn't be noticeable from an UX perspective.

Reply:
Although this could count towards the complex algorithm goal for the app :)